### PR TITLE
hw-06

### PR DESCRIPTION
**Pull request by vsevolod-prus**

### CODE REVIEW

File: hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Override
+    public Author update(Author author) {
+        return entityManager.merge(author);
+    }
+
+    @Override
+    public void updateFullNameById(Long id, String fullName) {
+        Query query = entityManager.createQuery("update Author a " +
```
Comment (kataus) [inline hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java]: А зачем это? У нас же есть ORM, пусть он для нас делает эту работу

File: hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                "set a.fullName = :fullName " +
+                "where a.id = :id");
+        query.setParameter("fullName", fullName);
+        query.setParameter("id", id);
+        query.executeUpdate();
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        Query query = entityManager.createQuery("delete from Author a where a.id = :id");
```
Comment (kataus) [inline hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java]: А чем em.remove плох? Такой подход достаточно опасен, потому что работает в обход кэша и сущность может там остаться, а из базы уже удалили

File: hw-06/src/main/java/ru/otus/service/impl/CommentsServiceJpa.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import ru.otus.model.Comment;
+import ru.otus.repository.api.CommentRepository;
+import ru.otus.service.api.CommentsService;
+
+import java.util.List;
+
+@Service
+@Transactional(readOnly = true)
```
Comment (kataus) [inline hw-06/src/main/java/ru/otus/service/impl/CommentsServiceJpa.java]: На занятии говорили почему тут не стоит размещать транзакцию. Да и она правда нам лишняя нужна регулярно?

File: hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java (hunk 1 above)
Comment (vsevolod-prus) [reply hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java]: Тут мы обновляем только колонку fullName. Для обновления всех полей у меня есть метод update.

File: hw-06/src/main/java/ru/otus/service/impl/CommentsServiceJpa.java (hunk 3 above)
Comment (vsevolod-prus) [reply hw-06/src/main/java/ru/otus/service/impl/CommentsServiceJpa.java]: На занятиях говорили лишь об том, чтоб не ставить на сервисы @transactional, но причину не называли. В некоторых источниках пишут что так можно и нужно делать, т.к. мы по умолчанию делаем все методы transactional(readonly=true). Если нужно переопределять поведение - переопределяем. (https://vladmihalcea.com/spring-transactional-annotation/ к примеру тут). По поводу лишней транзакции не понял - https://stackoverflow.com/questions/46090335/should-i-use-transactional-for-simple-queries-with-jdbctemplate

File: hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java (hunk 2 above)
Comment (vsevolod-prus) [reply hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java]: Можем вызывать где надо em.clear()/em.detach(). Если же мы знаем что у нас объект лежит в контексте, мы можем написать тут метод delete(Author author){em.remove(author)}, но такой потребности в моём ДЗ не было. Что если мы знаем только id объекта, а самого объекта у нас нет ? Вы предлагаете сделать em.find(id), достать весь объект, ради того чтобы его потом удалить по айдишке ? К слову, в примерах к лекциям удаление происходит именно через createQuery.

File: hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java (hunk 1 above)
Comment (kataus) [reply hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java]: Так не делается обычно, экономия копеечная, но убивает всё удобноство ORM

File: hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java (hunk 2 above)
Comment (kataus) [reply hw-06/src/main/java/ru/otus/repository/impl/AuthorRepositoryJpa.java]: И там говорилось - что одиночный надежнее через remove, на массовых уже имеет смысл делать query.
delete - самая дорогая операция на  DB, потому небольшой селект перед ней почти не повлияет на общий результат. + delete используем на самом деле очень редко. А вот вылавливать фокусы с фантомной записью в кэше будет весело

File: hw-06/src/main/java/ru/otus/service/impl/CommentsServiceJpa.java (hunk 3 above)
Comment (kataus) [reply hw-06/src/main/java/ru/otus/service/impl/CommentsServiceJpa.java]: 1. Потому что наверху в классе оно потеряется и забудем о том, что есть транзакционность. 
2. Не стоит тянуть функционал, от которого нет пользы. К тому же достаточно дорогой. Это довольно плохая практика, отрицательно сказывающаяся на общей производительности и сопровождаемости
3.

File: hw-06/src/main/java/ru/otus/repository/impl/CommentRepositoryJpa.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+@Repository
+@RequiredArgsConstructor
+public class CommentRepositoryJpa implements CommentRepository {
+
+    @PersistenceContext
+    private final EntityManager entityManager;
+
+    @Override
+    public void updateTextById(Long id, String value) {
```
Comment (kataus) [inline hw-06/src/main/java/ru/otus/repository/impl/CommentRepositoryJpa.java:21]: А эта задача относится скорее к слою бизнес-логики, иначе у нас на dao начнет плодиться функционал, а это нарушает концепцию слоя
